Add LNS tutorial - Closes #1151 (#1158)
LiskHQ/lisk-docs: modules/ROOT/pages/tutorials/lns/2-frontend.adoc

= Chapter 2: Frontend
Mona Bärenfänger <mona@lightcurve.io>
// Settings
:toc: preamble
:toclevels: 4
:idprefix:
:idseparator: -
:imagesdir: ../../assets/images
:experimental:
// URLs
:url_sdk_dashboardplugin: https://github.com/LiskHQ/lisk-sdk/tree/v5.1.4/framework-plugins/lisk-framework-dashboard-plugin
:url_sdkexamples_lns_uiplugin: https://github.com/LiskHQ/lisk-sdk-examples/tree/development/tutorials/lisk-name-service/lns-dashboard-plugin
:url_reactjs: https://reactjs.org/docs/create-a-new-react-app.html#create-react-app
:url_reactjs_docs: https://reactjs.org/docs/create-a-new-react-app.html
// Project URLs
:url_lns_0: tutorials/lns/index.adoc#lns-application-overview
:url_lns_1_newacc: tutorials/lns/1-blockchain.adoc#create-new-account
:url_lns_1_register: tutorials/lns/1-blockchain.adoc#register-domain
:url_bapps_frontend: introduction/blockchain-applications.adoc#frontend-backend

== Using a plugin as frontend

The SDK Dashboard plugin is nice to use during development of th blockchain application, because it offers a simple way to interact with the blockchain application through a user interface.

> To make it more convenient for normal users to use the application, add a frontend to the application which is specialized on the respective use case of the LNS applicationn.

Author: ```suggestion
To make the application more user-friendly, add a frontend which is specialized for the respective use case of the LNS application.
```

> For the LNS app, we want to provide a simple *frontend as a plugin*, which is registered to the LNS blockchain application.

Author: it is not really recommended, it is more one of multiple approaches how to add a frontend to the application.

The development of the React application is not covered in this tutorial in detail.
Instead, we will use the existing React frontend in the {url_sdkexamples_lns_uiplugin}[lisk-sdk-examples repository^] and see how it can be included as a standalone UI plugin for a blockchain application.

TIP: If you want to learn more about how to develop a React frontend application, check out the {url_reactjs_docs}[React.js documentation^].

.lisk-name-service/
[source,bash]
----
cd .. # <1>
git clone https://github.com/LiskHQ/lisk-sdk-examples.git # <2>
cp -R lisk-sdk-examples/tutorials/lisk-name-service/lns-dashboard-plugin lisk-name-service/lns-dashboard-plugin # <3>
----

<1> Move out of the `lisk-name-service` folder.
<2> Clone the `lisk-sdk-examples` repository.
<3> Copy the LNS Dashboard plugin to the root folder of your LNS blockchain application.

As you might notice when looking at the files, the LNS Dashboard plugin is based on the code of the {url_sdk_dashboardplugin}[Dashboard plugin^] from the Lisk SDK Framework.

The file structure of the UI plugin is basically a merge of a React.js application and a Lisk plugin.

> The main logic of the React application is located under `lns-dashboard-plugin/src/ui/`, and the main logic fie the LNS UI Plugin is located in `lns-dashboard-plugin/src/app/`.

Author: ```suggestion
The main logic of the React application is located under `lns-dashboard-plugin/src/ui/`, and the main logic of the LNS UI Plugin is located in `lns-dashboard-plugin/src/app/`.
```

		port: 8000,
	});
};
----

Save and close `plugins.ts`.
Restart the LNS blockchain application to apply the changes.

.lisk-name-service/lns/
[source,bash]
----
./bin/run start
----

After the application has loaded, it is possible to acceess the LNS Dashboard under http://localhost:8000.

image:tutorials/lns/ui-disconnected.png[UI disconnected,200,100]

Use the credentials of the account that was created newly in the previous step xref:{url_lns_1_newacc}[Create new account] of Chapter 1 to connect to the LNS frontend.

image:tutorials/lns/ui-connected.png[UI connected,200,100]

At the top right corner you will now see the domain that was defined as <<define-a-reverse-lookup-address,reverse lookup>> for the account address.
If you tick of the slider `Enable LNS`, the address of the account will be displayed again, instead of the domain name.

By clicking on your account you will reach the following page, giving all important information about your account and the domain names that are registered for this account.
Currently, one domain name is registered to the account.
This was done in step xref:{url_lns_1_register}[Register Domain] of Chapter 1 via the Dashboard plugin.

> You can also see the new TXT record that was added to the LNS object.

Author: ```suggestion
The new TXT record, which was added to the LNS object, can also be seen.
```

You should see the new domain name listed there.

image:tutorials/lns/ui-account2.png[UI account page 2,200,100]
image:tutorials/lns/ui-account-awesome-details.png[UI domain details,200,100]

Unfortunately, on the top right, our address is still resolving to the `my-name.lsk` domain (if LNS is enabled).
So let's update the reverse lookup of the account to point to the new domain name `awesome.lsk`.

Open the Dialog for updating the reverse lookup by clicking on kbd:[Update reverse lookup] in the top right.

image:tutorials/lns/ui-reverse-lookup-dialog.png[UI reverse lookup dialog,200,100]

Choose `awesome.lsk` from the dropdown menu, enter the account passphrase and the minimum fee.

Click kbd:[Update] to update the reverse lookup entry for this account.
As a result, you can verify that the LNS Dashboard now resolves the account address automatically to the new domain name `awesome.lsk`.

image:tutorials/lns/ui-updated.png[UI updated name,200,100]

Try to search again for the domain name `awesome.lsk`.
The search results should now inform you that this domain name is already reserved.

image:tutorials/lns/ui-reach-reserved.png[UI search for reserved name,200,100]

> Play around as much as you like with the LNS application UI.

Author: ```suggestion
Familiarize yourself as much as you like with the LNS application UI before moving on.
```
